File: qubit_computer/gates.py
"""Gate library for Qubit Computer.

Gates are plain nested lists of ``complex`` (row-major unitary matrices)
so the simulator has no third-party dependencies.  Parametric gates are
functions returning a matrix; fixed gates are module-level constants.

The APQB-native gates are:

* ``APQB(theta)``   -- RY(2 theta): |0> -> cos(theta)|0> + sin(theta)|1>
* ``APQB_R(r)``     -- prepare from a correlation coefficient r = cos(2 theta)
* ``APQB_A(a)``     -- prepare from an unconstrained latent a (tanh / sech)
"""

from __future__ import annotations

import cmath
import math
from typing import Callable, Dict, List, Sequence

from .apqb import theta_from_latent, theta_from_r

Matrix = List[List[complex]]

_SQRT1_2 = 1.0 / math.sqrt(2.0)

# ----------------------------------------------------------------- 1-qubit
I: Matrix = [[1, 0], [0, 1]]
X: Matrix = [[0, 1], [1, 0]]
Y: Matrix = [[0, -1j], [1j, 0]]
Z: Matrix = [[1, 0], [0, -1]]
H: Matrix = [[_SQRT1_2, _SQRT1_2], [_SQRT1_2, -_SQRT1_2]]
S: Matrix = [[1, 0], [0, 1j]]
SDG: Matrix = [[1, 0], [0, -1j]]
T: Matrix = [[1, 0], [0, cmath.exp(1j * math.pi / 4)]]
TDG: Matrix = [[1, 0], [0, cmath.exp(-1j * math.pi / 4)]]
SX: Matrix = [[0.5 + 0.5j, 0.5 - 0.5j], [0.5 - 0.5j, 0.5 + 0.5j]]


def RX(theta: float) -> Matrix:
    c, s = math.cos(theta / 2), math.sin(theta / 2)
    return [[c, -1j * s], [-1j * s, c]]


def RY(theta: float) -> Matrix:
    c, s = math.cos(theta / 2), math.sin(theta / 2)
    return [[c, -s], [s, c]]


def RZ(theta: float) -> Matrix:
    return [[cmath.exp(-1j * theta / 2), 0], [0, cmath.exp(1j * theta / 2)]]


def P(lam: float) -> Matrix:
    """Phase gate diag(1, e^{i lam})."""
    return [[1, 0], [0, cmath.exp(1j * lam)]]


def U(theta: float, phi: float, lam: float) -> Matrix:
    """Generic single-qubit rotation U(theta, phi, lambda) (OpenQASM convention)."""
    c, s = math.cos(theta / 2), math.sin(theta / 2)
    return [
        [c, -cmath.exp(1j * lam) * s],
        [cmath.exp(1j * phi) * s, cmath.exp(1j * (phi + lam)) * c],
    ]


def APQB(theta: float) -> Matrix:
    """APQB preparation gate: RY(2 theta).

    Acting on |0> it yields the APQB state cos(theta)|0> + sin(theta)|1>,
    so r = cos(2 theta) and T = |sin(2 theta)| are read back exactly.
    """
    return RY(2.0 * theta)


def APQB_R(r: float) -> Matrix:
    """APQB preparation from a correlation coefficient r in [-1, 1]."""
    return APQB(theta_from_r(r))


def APQB_A(a: float) -> Matrix:
    """APQB preparation from an unconstrained latent a (r=tanh a, T=sech a)."""
    return APQB(theta_from_latent(a)[2])


# ----------------------------------------------------------------- 2-qubit
# Convention: for a k-qubit matrix the basis index is
#   idx = sum_j bit(target_j) << (k-1-j)
# i.e. the FIRST target listed is the most significant bit.  For CX this
# means targets=(control, target).


def controlled(u: Matrix) -> Matrix:
    """Return the controlled version of a k-qubit unitary (control = first qubit)."""
    n = len(u)
    dim = 2 * n
    m: Matrix = [[0j] * dim for _ in range(dim)]
    for i in range(n):
        m[i][i] = 1
    for i in range(n):
        for j in range(n):
            m[n + i][n + j] = u[i][j]
    return m


CX: Matrix = controlled(X)
CY: Matrix = controlled(Y)
CZ: Matrix = controlled(Z)
CH: Matrix = controlled(H)
SWAP: Matrix = [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 0, 1]]
ISWAP: Matrix = [[1, 0, 0, 0], [0, 0, 1j, 0], [0, 1j, 0, 0], [0, 0, 0, 1]]


def CRX(theta: float) -> Matrix:
    return controlled(RX(theta))


def CRY(theta: float) -> Matrix:
    return controlled(RY(theta))


def CRZ(theta: float) -> Matrix:
    return controlled(RZ(theta))


def CP(lam: float) -> Matrix:
    return controlled(P(lam))


def CAPQB(theta: float) -> Matrix:
    """Controlled APQB preparation (controlled RY(2 theta))."""
    return controlled(APQB(theta))


def RXX(theta: float) -> Matrix:
    c, s = math.cos(theta / 2), -1j * math.sin(theta / 2)
    return [[c, 0, 0, s], [0, c, s, 0], [0, s, c, 0], [s, 0, 0, c]]


def RZZ(theta: float) -> Matrix:
    e_m, e_p = cmath.exp(-1j * theta / 2), cmath.exp(1j * theta / 2)
    return [[e_m, 0, 0, 0], [0, e_p, 0, 0], [0, 0, e_p, 0], [0, 0, 0, e_m]]


# ----------------------------------------------------------------- 3-qubit
CCX: Matrix = controlled(CX)  # Toffoli, targets=(c1, c2, t)
CSWAP: Matrix = controlled(SWAP)  # Fredkin, targets=(c, a, b)


# ----------------------------------------------------------------- registry
FIXED_GATES: Dict[str, Matrix] = {
    "i": I, "id": I, "x": X, "y": Y, "z": Z, "h": H, "s": S, "sdg": SDG,
    "t": T, "tdg": TDG, "sx": SX,
    "cx": CX, "cnot": CX, "cy": CY, "cz": CZ, "ch": CH, "swap": SWAP,
    "iswap": ISWAP, "ccx": CCX, "toffoli": CCX, "cswap": CSWAP, "fredkin": CSWAP,
}

PARAM_GATES: Dict[str, Callable[..., Matrix]] = {
    "rx": RX, "ry": RY, "rz": RZ, "p": P, "u": U,
    "apqb": APQB, "apqb_r": APQB_R, "apqb_a": APQB_A,
    "crx": CRX, "cry": CRY, "crz": CRZ, "cp": CP, "capqb": CAPQB,
    "rxx": RXX, "rzz": RZZ,
}

GATE_ARITY: Dict[str, int] = {
    **{k: 1 for k in ("i", "id", "x", "y", "z", "h", "s", "sdg", "t", "tdg", "sx",
                      "rx", "ry", "rz", "p", "u", "apqb", "apqb_r", "apqb_a")},
    **{k: 2 for k in ("cx", "cnot", "cy", "cz", "ch", "swap", "iswap",
                      "crx", "cry", "crz", "cp", "capqb", "rxx", "rzz")},
    **{k: 3 for k in ("ccx", "toffoli", "cswap", "fredkin")},
}


def resolve(name: str, params: Sequence[float] = ()) -> Matrix:
    """Look up a gate by name and instantiate it with ``params``."""
    key = name.lower()
    if key in FIXED_GATES:
        if params:
            raise ValueError(f"gate '{name}' takes no parameters")
        return FIXED_GATES[key]
    if key in PARAM_GATES:
        return PARAM_GATES[key](*params)
    raise KeyError(f"unknown gate '{name}'")


def is_unitary(m: Matrix, tol: float = 1e-9) -> bool:
    """Check m^dagger m = I."""
    n = len(m)
    for i in range(n):
        for j in range(n):
            acc = 0j
            for k in range(n):
                acc += m[k][i].conjugate() * m[k][j]
            if abs(acc - (1.0 if i == j else 0.0)) > tol:
                return False
    return True


def dagger(m: Matrix) -> Matrix:
    n = len(m)
    return [[complex(m[j][i]).conjugate() for j in range(n)] for i in range(n)]

File: app/src/core/gates.ts
/** Gate library. Matrices are row-major arrays of Complex. */
import { C, Complex, cexp, cmul, conj } from './complex';
import { thetaFromLatent, thetaFromR } from './apqb';

export type Matrix = Complex[][];

const S2 = 1 / Math.SQRT2;
const c = C;

export const I: Matrix = [[c(1), c(0)], [c(0), c(1)]];
export const X: Matrix = [[c(0), c(1)], [c(1), c(0)]];
export const Y: Matrix = [[c(0), c(0, -1)], [c(0, 1), c(0)]];
export const Z: Matrix = [[c(1), c(0)], [c(0), c(-1)]];
export const H: Matrix = [[c(S2), c(S2)], [c(S2), c(-S2)]];
export const S: Matrix = [[c(1), c(0)], [c(0), c(0, 1)]];
export const SDG: Matrix = [[c(1), c(0)], [c(0), c(0, -1)]];
export const T: Matrix = [[c(1), c(0)], [c(0), cexp(Math.PI / 4)]];
export const TDG: Matrix = [[c(1), c(0)], [c(0), cexp(-Math.PI / 4)]];
export const SX: Matrix = [[c(0.5, 0.5), c(0.5, -0.5)], [c(0.5, -0.5), c(0.5, 0.5)]];

export const RX = (t: number): Matrix => {
  const co = Math.cos(t / 2);
  const si = Math.sin(t / 2);
  return [[c(co), c(0, -si)], [c(0, -si), c(co)]];
};
export const RY = (t: number): Matrix => {
  const co = Math.cos(t / 2);
  const si = Math.sin(t / 2);
  return [[c(co), c(-si)], [c(si), c(co)]];
};
export const RZ = (t: number): Matrix => [[cexp(-t / 2), c(0)], [c(0), cexp(t / 2)]];
export const P = (lam: number): Matrix => [[c(1), c(0)], [c(0), cexp(lam)]];
export const U = (t: number, phi: number, lam: number): Matrix => {
  const co = Math.cos(t / 2);
  const si = Math.sin(t / 2);
  return [
    [c(co), cmul(cexp(lam), c(-si))],
    [cmul(cexp(phi), c(si)), cmul(cexp(phi + lam), c(co))],
  ];
};

/** APQB preparation gate RY(2 theta): |0> -> cos(theta)|0> + sin(theta)|1>. */
export const APQB = (theta: number): Matrix => RY(2 * theta);
export const APQB_R = (r: number): Matrix => APQB(thetaFromR(r));
export const APQB_A = (a: number): Matrix => APQB(thetaFromLatent(a)[2]);

/** Controlled version of a k-qubit unitary (control = first target). */
export function controlled(u: Matrix): Matrix {
  const n = u.length;
  const dim = 2 * n;
  const m: Matrix = [];
  for (let i = 0; i < dim; i++) {
    const row: Complex[] = [];
    for (let j = 0; j < dim; j++) row.push(c(0));
    m.push(row);
  }
  for (let i = 0; i < n; i++) m[i][i] = c(1);
  for (let i = 0; i < n; i++) for (let j = 0; j < n; j++) m[n + i][n + j] = u[i][j];
  return m;
}

export const CX = controlled(X);
export const CY = controlled(Y);
export const CZ = controlled(Z);
export const CH = controlled(H);
export const SWAP: Matrix = [
  [c(1), c(0), c(0), c(0)],
  [c(0), c(0), c(1), c(0)],
  [c(0), c(1), c(0), c(0)],
  [c(0), c(0), c(0), c(1)],
];
export const ISWAP: Matrix = [
  [c(1), c(0), c(0), c(0)],
  [c(0), c(0), c(0, 1), c(0)],
  [c(0), c(0, 1), c(0), c(0)],
  [c(0), c(0), c(0), c(1)],
];
export const CRX = (t: number): Matrix => controlled(RX(t));
export const CRY = (t: number): Matrix => controlled(RY(t));
export const CRZ = (t: number): Matrix => controlled(RZ(t));
export const CP = (l: number): Matrix => controlled(P(l));
export const CAPQB = (t: number): Matrix => controlled(APQB(t));
export const RXX = (t: number): Matrix => {
  const co = c(Math.cos(t / 2));
  const si = c(0, -Math.sin(t / 2));
  return [[co, c(0), c(0), si], [c(0), co, si, c(0)], [c(0), si, co, c(0)], [si, c(0), c(0), co]];
};
export const RZZ = (t: number): Matrix => {
  const em = cexp(-t / 2);
  const ep = cexp(t / 2);
  return [[em, c(0), c(0), c(0)], [c(0), ep, c(0), c(0)], [c(0), c(0), ep, c(0)], [c(0), c(0), c(0), em]];
};
export const CCX = controlled(CX);
export const CSWAP = controlled(SWAP);

export const FIXED_GATES: Record<string, Matrix> = {
  i: I, id: I, x: X, y: Y, z: Z, h: H, s: S, sdg: SDG, t: T, tdg: TDG, sx: SX,
  cx: CX, cnot: CX, cy: CY, cz: CZ, ch: CH, swap: SWAP, iswap: ISWAP,
  ccx: CCX, toffoli: CCX, cswap: CSWAP, fredkin: CSWAP,
};

export const PARAM_GATES: Record<string, (...p: number[]) => Matrix> = {
  rx: RX, ry: RY, rz: RZ, p: P, u: U, apqb: APQB, apqb_r: APQB_R, apqb_a: APQB_A,
  crx: CRX, cry: CRY, crz: CRZ, cp: CP, capqb: CAPQB, rxx: RXX, rzz: RZZ,
};

export const GATE_ARITY: Record<string, number> = {};
for (const k of ['i', 'id', 'x', 'y', 'z', 'h', 's', 'sdg', 't', 'tdg', 'sx', 'rx', 'ry', 'rz', 'p', 'u', 'apqb', 'apqb_r', 'apqb_a']) GATE_ARITY[k] = 1;
for (const k of ['cx', 'cnot', 'cy', 'cz', 'ch', 'swap', 'iswap', 'crx', 'cry', 'crz', 'cp', 'capqb', 'rxx', 'rzz']) GATE_ARITY[k] = 2;
for (const k of ['ccx', 'toffoli', 'cswap', 'fredkin']) GATE_ARITY[k] = 3;

export const GATE_PARAM_COUNT: Record<string, number> = { u: 3 };
for (const k of Object.keys(PARAM_GATES)) if (!(k in GATE_PARAM_COUNT)) GATE_PARAM_COUNT[k] = 1;

export function resolve(name: string, params: number[] = []): Matrix {
  const key = name.toLowerCase();
  if (key in FIXED_GATES) {
    if (params.length) throw new Error(`gate '${name}' takes no parameters`);
    return FIXED_GATES[key];
  }
  if (key in PARAM_GATES) {
    const need = GATE_PARAM_COUNT[key];
    if (params.length !== need) throw new Error(`gate '${name}' needs ${need} parameter(s)`);
    return PARAM_GATES[key](...params);
  }
  throw new Error(`unknown gate '${name}'`);
}

export function isUnitary(m: Matrix, tol = 1e-9): boolean {
  const n = m.length;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      let re = 0;
      let im = 0;
      for (let k = 0; k < n; k++) {
        const p = cmul(conj(m[k][i]), m[k][j]);
        re += p.re;
        im += p.im;
      }
      if (Math.abs(re - (i === j ? 1 : 0)) > tol || Math.abs(im) > tol) return false;
    }
  }
  return true;
}
